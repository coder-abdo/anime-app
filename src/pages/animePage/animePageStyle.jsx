import styled from "styled-components";

export const PageContainer = styled.div`
  display: flex;
  flex-direction: column;
  p {
    font-size: 1.6rem;
    margin: 2rem 0 3rem;
    color: var(--secondary-color);
    font-weight: 400;
  }
`;
export const Image = styled.img`
  width: 50rem;
  height: 30rem;
  object-fit: cover;
  align-self: center;
`;
