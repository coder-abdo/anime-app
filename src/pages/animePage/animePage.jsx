import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import {
  fetchSingleAnime,
  fetchCharacters
} from "../../redux/actions/fetchAnimes";
import { Loader } from "../../components/loader/Loader";
import { Character } from "../../components/character/Character";
import {
  AnimesTitle,
  Container,
  ButtonLink
} from "../../sharedStyle/Container";
import { Image, PageContainer } from "./animePageStyle";
export const AnimePage = () => {
  const { id } = useParams();
  const { singleAnime, characters } = useSelector(({ animes }) => animes);
  // console.log(singleAnime, characters);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchSingleAnime(id));
    dispatch(fetchCharacters(id));
    // eslint-disable-next-line
  }, []);
  return (
    <>
      {singleAnime ? (
        <PageContainer>
          <AnimesTitle data-testid="single-anime-title">
            {singleAnime.title}
          </AnimesTitle>
          <Image src={singleAnime.image_url} alt={singleAnime.title} />
          <p data-testid="synopsis">{singleAnime.synopsis}</p>
          {!!characters.length ? (
            <Container data-testid="characters-container">
              {characters
                .sort((a, b) => +a.male_id - +b.male_id)
                .map(character => (
                  <Character key={character.mal_id} {...character} />
                ))}
            </Container>
          ) : (
            <Loader />
          )}
          <ButtonLink isback="back" to="/" data-testid="back-btn">
            Back to Animes
          </ButtonLink>
        </PageContainer>
      ) : (
        <Loader />
      )}
    </>
  );
};
