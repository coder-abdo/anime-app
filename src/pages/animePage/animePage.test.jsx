import React from "react";
import { Provider } from "react-redux";
import {
  render,
  cleanup,
  fireEvent,
  waitForElement
} from "@testing-library/react";
import { Router } from "react-router";
import { createBrowserHistory } from "history";
import "@testing-library/jest-dom";
import { Store } from "../../redux/store/store";
import { AnimePage } from "./animePage";
afterEach(cleanup);

describe("render single anime page", () => {
  const history = createBrowserHistory();
  const { container, asFragment } = render(
    <Provider store={Store}>
      <Router history={history}>
        <AnimePage />
      </Router>
    </Provider>
  );
  test("the anime page render in the document", () => {
    expect(container).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  test("should render loader before characters of anime", () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <AnimePage />
        </Router>
      </Provider>
    );
    expect(getByTestId("loader")).toBeInTheDocument();
  });
  test("should render the characters of anime after some delay", async () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <AnimePage />
        </Router>
      </Provider>
    );
    const charactersList = await waitForElement(() =>
      getByTestId("characters-container")
    );
    expect(charactersList).toBeInTheDocument();
  });
  test("should go back to home page after clicking the submit button", () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <AnimePage />
        </Router>
      </Provider>
    );
    fireEvent.click(getByTestId("back-btn"));
    console.log(history.length);
    expect(history.length).toBe(2);
  });
});
