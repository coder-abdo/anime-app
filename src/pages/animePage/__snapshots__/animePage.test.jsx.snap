// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render single anime page the anime page render in the document 1`] = `
<DocumentFragment>
  <div
    class="sc-AxheI clsAXF"
  >
    <h1
      class="sc-AxhUy iWKeEK"
      data-testid="single-anime-title"
    />
    <img
      class="sc-Axmtr juqOmg"
    />
    <p
      data-testid="synopsis"
    />
    <div
      class="sc-AxjAm efKchL"
      data-testid="loader"
    >
      <div
        class="sc-AxirZ dZCFLP"
      />
    </div>
    <a
      class="sc-AxgMl cbKzbU"
      data-testid="back-btn"
      href="/"
      isback="back"
    >
      Back to Animes
    </a>
  </div>
</DocumentFragment>
`;
